File: src/graph/components.py
"""連結成分抽出モジュール"""

import numpy as np
import scipy.sparse as sp
import scipy.sparse.csgraph as csgraph
from src.types import OUTLIER, Int32Array


def extract_components(adj: sp.csr_matrix) -> Int32Array:
    """連結成分を抽出

    Args:
        adj: 隣接行列

    Returns:
        Int32Array: 成分ラベル (n,)
    """
    # Definition 9: connected components of the mutual k-NN graph.
    _, labels = csgraph.connected_components(adj, directed=False, return_labels=True)
    return labels.astype(np.int32)


def filter_by_component_size(
    components: Int32Array,
    min_size: int,
) -> Int32Array:
    """小さな成分を外れ値としてマーク

    Args:
        components: 成分ラベル
        min_size: 最小成分サイズ（これ以下は外れ値）

    Returns:
        Int32Array: 外れ値をOUTLIER(-1)でマークした成分ラベル
    """
    result = components.copy()
    labels, counts = np.unique(components, return_counts=True)
    small_components = labels[counts <= min_size]
    mask = np.isin(components, small_components)
    result[mask] = OUTLIER
    return result


def filter_by_edge_count(
    adj: sp.csr_matrix,
    components: Int32Array,
    min_edges: int,
) -> Int32Array:
    """辺数が少ない点を外れ値としてマーク（論文準拠）

    Args:
        adj: 隣接行列
        components: 成分ラベル
        min_edges: 最小辺数（これ以下は外れ値）

    Returns:
        Int32Array: 外れ値をOUTLIER(-1)でマークした成分ラベル
    """
    result = components.copy()
    degrees = np.array(adj.getnnz(axis=1)).flatten()
    mask = degrees <= min_edges
    result[mask] = OUTLIER
    return result

File: rust/src/graph/components.rs
//! Connected components and outlier filtering.

use crate::graph::adjacency::WeightedGraph;
use crate::types::OUTLIER;

#[derive(Debug, Clone, Copy)]
pub enum OutlierFilter {
    None,
    EdgeCount,
    ComponentSize,
}

pub fn extract_components(graph: &WeightedGraph) -> Vec<i32> {
    let n = graph.n();
    let mut labels = vec![-1i32; n];
    let mut comp_id = 0i32;

    let mut stack: Vec<usize> = Vec::new();
    for start in 0..n {
        if labels[start] != -1 {
            continue;
        }
        labels[start] = comp_id;
        stack.push(start);

        while let Some(v) = stack.pop() {
            for &(u, _) in &graph.adj[v] {
                if labels[u] == -1 {
                    labels[u] = comp_id;
                    stack.push(u);
                }
            }
        }

        comp_id += 1;
    }

    labels
}

pub fn filter_by_edge_count(
    graph: &WeightedGraph,
    components: &[i32],
    min_edges: usize,
) -> Vec<i32> {
    let n = graph.n();
    let mut out = components.to_vec();
    for i in 0..n {
        if graph.degree(i) <= min_edges {
            out[i] = OUTLIER;
        }
    }
    out
}

pub fn filter_by_component_size(components: &[i32], min_size: usize) -> Vec<i32> {
    let mut counts: std::collections::HashMap<i32, usize> = std::collections::HashMap::new();
    for &c in components {
        *counts.entry(c).or_insert(0) += 1;
    }

    let mut out = components.to_vec();
    for (i, &c) in components.iter().enumerate() {
        if let Some(&sz) = counts.get(&c) {
            if sz <= min_size {
                out[i] = OUTLIER;
            }
        }
    }

    out
}

pub fn apply_outlier_filter(
    graph: &WeightedGraph,
    components: &[i32],
    filter: OutlierFilter,
    cutoff: usize,
) -> Vec<i32> {
    match filter {
        OutlierFilter::None => components.to_vec(),
        OutlierFilter::EdgeCount => filter_by_edge_count(graph, components, cutoff),
        OutlierFilter::ComponentSize => filter_by_component_size(components, cutoff),
    }
}
